### PR TITLE
taco integration

### PR DESCRIPTION
**Pull request by huangjd**

### CODE REVIEW

File: src/lower/lower_indexexpr_taco.cpp [hunk 1]
```diff
@@ -0,0 +1,1168 @@
...
+      });
+  if (sparsity.size() > 0) {
+    sparsity[0] = false;
+  }
+  return sparsity;
+}
+
+// The same index var appears on different dimensions at different expressions,
+// while the sparsities of these dimensions do not agree
+// For example  A(i,j) = B(i,k)*B(k,j), where B is {Dense, Sparse}
```
Comment (stephenchouca) [inline src/lower/lower_indexexpr_taco.cpp]: Is this actually an issue? taco seems to be able to generate code for the particular example in the comment.

File: src/lower/lower_indexexpr_taco.cpp [hunk 2]
```diff
@@ -0,0 +1,1168 @@
...
+            "diagonal or stencil storage");
+      CHECK(hasFixedIndexVar(stmt), "fixed index expression");
+      CHECK(hasDivision(stmt), "division");
+      return false;
+    }
+  } rewriter;
+
+  func = rewriter.rewrite(func);
+  func = insertVarDecls(func);
+  func = insertInitializations(func); /// @todo needed due to bugs in taco
```
Comment (stephenchouca) [inline src/lower/lower_indexexpr_taco.cpp]: Is this still needed with the latest version of taco? I believe it should now generate output initialization code as well.

File: src/lower/lower_indexexpr_taco.cpp [hunk 3]
```diff
@@ -0,0 +1,1168 @@
...
+  virtual void visit(const Sub* op) {
+    Ptr a = compile(op->a);
+    Ptr b = compile(op->b);
+    tacoAccess = Ptr(new taco::IndexExpr(*a - *b));
+  }
+
+  virtual void visit(const Mul* op) {
+    Ptr a = compile(op->a);
+    Ptr b = compile(op->b);
+    tacoAccess = Ptr(new taco::IndexExpr(*a * *b));
```
Comment (stephenchouca) [inline src/lower/lower_indexexpr_taco.cpp]: Enclose `*a` and `*b` with parentheses to avoid confusion.

File: src/lower/lower_indexexpr_taco.cpp [hunk 4]
```diff
@@ -0,0 +1,1168 @@
...
+  virtual void visit(const Add* op) {
+    Ptr a = compile(op->a);
+    Ptr b = compile(op->b);
+    tacoAccess = Ptr(new taco::IndexExpr(*a + *b));
+  }
+
+  virtual void visit(const Sub* op) {
+    Ptr a = compile(op->a);
+    Ptr b = compile(op->b);
+    tacoAccess = Ptr(new taco::IndexExpr(*a - *b));
```
Comment (stephenchouca) [inline src/lower/lower_indexexpr_taco.cpp]: Enclose `*a` and `*b` with parentheses to avoid confusion.

File: src/lower/lower_indexexpr_taco.cpp [hunk 5]
```diff
@@ -0,0 +1,1168 @@
...
+
+  virtual void visit(const Neg* op) {
+    Ptr a = compile(op->a);
+    tacoAccess = Ptr(new taco::IndexExpr(-*a));
+  }
+
+  virtual void visit(const Add* op) {
+    Ptr a = compile(op->a);
+    Ptr b = compile(op->b);
+    tacoAccess = Ptr(new taco::IndexExpr(*a + *b));
```
Comment (stephenchouca) [inline src/lower/lower_indexexpr_taco.cpp]: Enclose `*a` and `*b` with parentheses to avoid confusion.

File: src/lower/lower_indexexpr_taco.cpp [hunk 6]
```diff
@@ -0,0 +1,1168 @@
...
+    tacoAccess = Ptr(new taco::Access(pair.first(pair.second)));
+  }
+
+  virtual void visit(const IndexExpr* op) {
+    tacoAccess = compile(op->value);
+  }
+
+  virtual void visit(const Neg* op) {
+    Ptr a = compile(op->a);
+    tacoAccess = Ptr(new taco::IndexExpr(-*a));
```
Comment (stephenchouca) [inline src/lower/lower_indexexpr_taco.cpp]: Enclose `*a` with parentheses to avoid confusion.

File: src/lower/lower_indexexpr_taco.cpp [hunk 7]
```diff
@@ -0,0 +1,1168 @@
...
+  virtual void visit(const Mul* op) {
+    Ptr a = compile(op->a);
+    Ptr b = compile(op->b);
+    tacoAccess = Ptr(new taco::IndexExpr(*a * *b));
+  }
+
+  virtual void visit(const Div* op) {
+    Ptr a = compile(op->a);
+    Ptr b = compile(op->b);
+    tacoAccess = Ptr(new taco::IndexExpr(*a / *b));;
```
Comment (stephenchouca) [inline src/lower/lower_indexexpr_taco.cpp]: Enclose `*a` and `*b` with parentheses to avoid confusion.

**Review by stephenchouca — CHANGES_REQUESTED** [review]

**Comment by huangjd** [comment]
Taco should throw exception on error, instead of calling abort(), since there's no way to recover from that

**Comment by huangjd** [comment]
Taco should throw exception on error instead of calling abort(), since there's no way to recover from that

**Pull request merged**